src/nanoarrow/integration/c_data_integration.h

```diff
@@ -20,10 +20,20 @@
 
 #include <stdint.h>
 
-#if defined(_MSC_VER)
-#define DLL_EXPORT __declspec(dllexport)
+#if defined _WIN32 || defined __CYGWIN__
+#define NANOARROW_DLL_IMPORT __declspec(dllimport)
+#define NANOARROW_DLL_EXPORT __declspec(dllexport)
+#define NANOARROW_DLL_LOCAL
 #else
-#define DLL_EXPORT
+#if __GNUC__ >= 4
+#define NANOARROW_DLL_IMPORT __attribute__((visibility("default")))
+#define NANOARROW_DLL_EXPORT __attribute__((visibility("default")))
+#define NANOARROW_DLL_LOCAL __attribute__((visibility("hidden")))
+#else
+#define NANOARROW_DLL_IMPORT
+#define NANOARROW_DLL_EXPORT
+#define NANOARROW_DLL_LOCAL
```

Comment:
Sorry about that - misinterpreted your earlier comment as wanting to go back to just the first few commits.

FWIW I looked for a prior art in some other entries in the Meson WrapDB. Here's one for sqlitecpp:

https://github.com/mesonbuild/wrapdb/blob/bd3e5f43c5b4e8ebc85ea8753407c07004c8159f/subprojects/packagefiles/sqlitecpp/meson.build#L220

docopt:

https://github.com/mesonbuild/wrapdb/blob/bd3e5f43c5b4e8ebc85ea8753407c07004c8159f/subprojects/packagefiles/docopt/meson.build#L7

luajit:

https://github.com/mesonbuild/wrapdb/blob/bd3e5f43c5b4e8ebc85ea8753407c07004c8159f/subprojects/packagefiles/luajit/src/meson.build#L191

I think the sqlitecpp approach is maybe the most relevant for us, i.e. we can try something like:

```python
if host_machine.system() == 'windows' and get_option('default_library') == 'shared'
    nanoarrow_args = ['-DNANOARROW_BUILD_DLL', '-DNANOARROW_EXPORT_DLL']
    nanoarrow_dep_args = ['-DNANOARROW_BUILD_DLL']
endif
```

Internally we use `nanoarrow_args` but third parties that use the dependency would automatically use the `nanoarrow_dep_args`. I think this approach is close to what you suggested, but instead of requiring consumers to provide `-DNANOARROW_CONSUME_DLL` we would internally use `-DNANOARROW_EXPORT_DLL` instead